File: 0057. Insert Interval/main0057.java
import java.util.Arrays;

public class main0057 {
    public static void main(String[] args) {
        Solution0057 sol = new Solution0057();
        int intervals[][], newInterval[], res[][];
        intervals = new int[][]{{1,3},{6,9}};
        newInterval = new int[]{2,5};
        res = sol.insert(intervals, newInterval);
        System.out.println(Arrays.deepToString(res));
    }
}


class Solution0057 {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        int[][] res = new int[intervals.length + 1][2];
        int idx = 0;
        // 遍历区间列表：
        // 首先将新区间左边且相离的区间加入结果集
        int i = 0;
        while (i < intervals.length && intervals[i][1] < newInterval[0]) {
            res[idx++] = intervals[i++];
        }
        // 接着判断当前区间是否与新区间重叠，重叠的话就进行合并，直到遍历到当前区间在新区间的右边且相离，
        // 将最终合并后的新区间加入结果集
        while (i < intervals.length && intervals[i][0] <= newInterval[1]) {
            newInterval[0] = Math.min(intervals[i][0], newInterval[0]);
            newInterval[1] = Math.max(intervals[i][1], newInterval[1]);
            i++;
        }
        res[idx++] = newInterval;
        // 最后将新区间右边且相离的区间加入结果集
        while (i < intervals.length) {
            res[idx++] = intervals[i++];
        }
        return Arrays.copyOf(res, idx);
    }


}
// https://leetcode-cn.com/problems/insert-interval/solution/bi-xu-miao-dong-li-kou-qu-jian-ti-mu-zhong-die-qu-/

File: 0057. Insert Interval/main0057.py
from typing import List


class Solution0057:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        n = len(intervals)
        idx, ridx = 0, 0
        output = []
        
        while idx < n and intervals[idx][1] < newInterval[0]:
            output.append(intervals[idx])
            idx += 1
        
        while idx < n and intervals[idx][0] <= newInterval[1]:
            newInterval[0] = min(newInterval[0], intervals[idx][0])            
            newInterval[1] = max(newInterval[1], intervals[idx][1])
            idx += 1
        output.append(newInterval)
        
        while idx < n and intervals[idx][0] > newInterval[1]:
            output.append(intervals[idx])
            idx += 1     
        
        return output

sol = Solution0057()
intervals = [[1,3],[6,9]]
newInterval = [2,5]
res = sol.insert(intervals,newInterval)
print(res)

intervals = [[1,2],[3,5],[6,7],[8,10],[12,16]]
newInterval = [4,8]
res = sol.insert(intervals,newInterval)
print(res)